proposals/0000-hyperdb.md

```diff
@@ -0,0 +1,581 @@
+0x05, 0x6C, 0x63, 0x9D, 0x87, 0xCA]`, which converts into the array:
+
+    [ 0, 3, 2, 2, 0, 3, 1, 3, 1, 1, 0, 0, 0, 3, 2, 1, 3, 0, 2, 1, 1, 3, 1, 2, 3, 1, 0, 2, 2, 2, 0, 3 ]
+
+
+`willow` has a 64-bit hash `[0x72, 0x30, 0x34, 0x39, 0x35, 0xA8, 0x21, 0x44]`,
+which converts into the array:
+
+    [ 2, 0, 3, 1, 0, 0, 3, 0, 0, 1, 3, 0, 1, 2, 3, 0, 1, 1, 3, 0, 0, 2, 2, 2, 1, 0, 2, 0, 0, 1, 0, 1 ]
+
+These combine into the unified byte array with 65 elements:
+
+    [ 0, 3, 2, 2, 0, 3, 1, 3, 1, 1, 0, 0, 0, 3, 2, 1, 3, 0, 2, 1, 1, 3, 1, 2, 3, 1, 0, 2, 2, 2, 0, 3,
+      2, 0, 3, 1, 0, 0, 3, 0, 0, 1, 3, 0, 1, 2, 3, 0, 1, 1, 3, 0, 0, 2, 2, 2, 1, 0, 2, 0, 0, 1, 0, 1,
+      4 ]
+
+As another example, the key `/a/b/c` converts into the 97-byte hash array:
+
+    [ 1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+      0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 3, 0, 3, 1, 3, 0, 1, 0, 1, 3, 2, 0, 2, 2, 3, 2, 2, 3, 3, 2, 3,
+      0, 1, 1, 0, 1, 2, 3, 2, 2, 2, 0, 0, 3, 1, 2, 1, 3, 3, 3, 3, 3, 3, 0, 3, 3, 2, 3, 2, 3, 0, 1, 0,
+      4 ]
+
+<!---
+
+Generation code (javascript) for the above:
+
+    var sodium = require('sodium-universal')
+    var toBuffer = require('to-buffer')
+
+    var KEY = Buffer.alloc(16)
+    var OUT = Buffer.alloc(8)
+
+    sodium.crypto_shorthash(OUT, toBuffer('tree'), KEY)
+    console.log("tree: ", OUT)
+    console.log(hash('tree', true))
+
+    sodium.crypto_shorthash(OUT, toBuffer('willow'), KEY)
+    console.log("willow: ", OUT)
+    console.log(hash('willow', true))
+
+    sodium.crypto_shorthash(OUT, toBuffer('a'), KEY)
+    console.log("a: ", OUT)
+    console.log(hash('a', true))
+
+Then, to manually "expand" arrays in python3:
+
+    hash_array = [0x72, 0x30, 0x34, 0x39, 0x35, 0xA8, 0x21, 0x44]
+    path = []
+    tmp = [(x & 0x3, (x >> 2) & 0x3, (x >> 4) & 0x3, (x >> 6) & 0x3) for x in hash_array]
+    [path.extend(e) for e in tmp]
+    path
+
+--->
+
+
+## Incremental Index Trie
+[trie_index]: #trie_index
+
+Each node stores a *prefix [trie](https://en.wikipedia.org/wiki/Trie)* that
+can be used to look up other keys, or to list all keys with a given prefix.
+This is stored under the `trie` field of the Node protobuf message.
+
+The trie effectively mirrors the `path` hash array. Each element in the `trie`
+points to the newest Node which has an identical path up to that specific
+prefix location. Elements can be null; any trailing null elements can be left
+blank.
+
+The data structure of the trie is a sparse array of pointers to other Node
+entries. Each pointer indicates a feed index and a sequence number; for the
+non-multi-writer case, the feed index is always 0, so we consider only the
+sequence number (aka, entry index).
+
+To lookup a key in the database, the recipe is to:
+
+1. Calculate the `path` array for the key you are looking for.
+2. Select the most-recent ("latest") Node for the feed.
+3. Compare path arrays. If the paths match exactly, you have found the Node you
```

Comment:
SipHash is not collision resistant, so two distinct paths could have the same path array, right?

Reply:
Yes, in case of a collision the final node bucket will just have more than one node in it. The full final node bucket is read on every read (usually only a single node) to do key full key comparison to check for collisions